src/mono/mono/mini/interp/transform-simd.c

```diff
@@ -283,47 +304,49 @@ emit_common_simd_operations (TransformData *td, int id, int atype, int vector_si
 }
 
 static gboolean
-get_common_simd_info (MonoClass *vector_klass, MonoMethodSignature *csignature, MonoTypeEnum *atype, int *vector_size, int *arg_size, int *scalar_arg)
+get_common_simd_info (MonoClass *klass, MonoTypeEnum *atype, int *klass_size, int *arg_size)
 {
-	if (!m_class_is_simd_type (vector_klass) && csignature->param_count)
-		vector_klass = mono_class_from_mono_type_internal (csignature->params [0]);
-	if (!m_class_is_simd_type (vector_klass))
+	*klass_size = mono_class_value_size (klass, NULL);
+	if (!m_class_is_simd_type (klass))
 		return FALSE;
-
-	MonoType *arg_type = mono_class_get_context (vector_klass)->class_inst->type_argv [0];
-	if (!mono_type_is_primitive (arg_type))
-		return FALSE;
-	*atype = arg_type->type;
-	if (*atype == MONO_TYPE_BOOLEAN)
-		return FALSE;
-	*vector_size = mono_class_value_size (vector_klass, NULL);
-	g_assert (*vector_size == SIZEOF_V128);
-	if (arg_size)
+	if (mono_class_is_ginst (klass)) {
+		MonoType *arg_type = mono_class_get_context (klass)->class_inst->type_argv [0];
+		if (!mono_type_is_primitive (arg_type))
+			return FALSE;
+		*atype = arg_type->type;
 		*arg_size = mono_class_value_size (mono_class_from_mono_type_internal (arg_type), NULL);
+		if (*atype == MONO_TYPE_BOOLEAN)
+			return FALSE;
+	} else {
+		*atype = MONO_TYPE_R4;
+		*arg_size = sizeof (float);
+	}
+	return TRUE;
+}
 
-	*scalar_arg = -1;
+static int get_common_simd_scalar_arg (MonoMethodSignature *csignature)
+{
 	for (int i = 0; i < csignature->param_count; i++) {
-		if (csignature->params [i]->type != MONO_TYPE_GENERICINST)
-			*scalar_arg = i;
+		if (MONO_TYPE_IS_PRIMITIVE(csignature->params [i]))
+			return i;
 	}
-
-	return TRUE;
+	return -1;
 }
 
 static void
-emit_common_simd_epilogue (TransformData *td, MonoClass *vector_klass, MonoMethodSignature *csignature, int vector_size, gboolean allow_void)
+emit_common_simd_epilogue (TransformData *td, MonoMethodSignature *csignature)
```

Comment:
To support methods which return a value type, but where that isn't a 128-bit vector, this explicitly gets the return type from the signature to ensure there can be no accidents